Replace manual LoggingContext usage with `ModuleApi.defer_to_threadpool`

anoadragon453 opened the pull request:
Attempt to replace manual usage of `LoggingContext` with the provided module API's `run_in_background` method.

I'm not entirely convinced about the changes to `fetch` (and subsequently `s3_download_task`). The fact we hand it a deferred directly is confusing me.

Spawning from https://github.com/matrix-org/synapse-s3-storage-provider/issues/133. https://github.com/matrix-org/synapse-s3-storage-provider/pull/74 can be closed after this PR is merged.

anoadragon453 committed:
Replace manual LoggingContext usage with `ModuleApi.run_in_background`

Attempt to replace manual usage of LoggingContext with the
provided module API's `run_in_background` method.

anoadragon453 commented on s3_storage_provider.py:
```diff
@@ -202,7 +185,7 @@ def parse_config(config):
 
 
-def s3_download_task(s3_client, bucket, key, extra_args, deferred, parent_logcontext):
+def s3_download_task(s3_client, bucket, key, extra_args, deferred):
```
The only real changes here are:

1. Remove `parent_logcontext`.
2. Removing `with LoggingContext ...` and de-indenting all of the code underneath it.

MadLittleMods replied:
I think these changes make sense. `s3_download_task` will use whatever the caller logcontext is.

And I think we maintain the logcontext when calling `s3_download_task` ✅ (at-least with the suggested patterns).

MadLittleMods commented on s3_storage_provider.py:
```diff
@@ -127,34 +115,29 @@ def _get_s3_client(self):
+                ExtraArgs=self.extra_args,
+            )
+
+        return run_in_background(
+            threads.deferToThreadPool, reactor, self._s3_pool, _store_file
```
While `run_in_background` will run `threads.deferToThreadPool` in the same logcontext, `threads.deferToThreadPool` will not run `_store_file` in any sort of logcontext.

What we really need is `synapse.logging.context.defer_to_threadpool(...)` which basically did the same pattern as before.

It looks like we do export this pattern via `ModuleApi.defer_to_thread(...)` which means we can also get rid of the `self._s3_pool`

```suggestion
        return module_api.defer_to_thread(reactor, _store_file)
```

anoadragon453 replied:
I've just realised we don't actually pass an instance of `ModuleApi` to `media_storage_providers`...

We could do so [here](https://github.com/element-hq/synapse/blob/50b74bac932def1f6970c82da3c09d38105b9a19/synapse/media/media_repository.py#L135), and honestly I see little reason not to. Not doing so hobbles what media storage providers can do.

MadLittleMods replied:
We could use `hs.get_module_api()`

anoadragon453 replied:
Actually, we can get it from `hs.get_module_api()`. Though I wonder whether we should take this as an opportunity to update the signature of `__init__` completely to only be `__init__(self, config, module_api)`?

MadLittleMods replied:
@anoadragon453 I haven't put enough thought into it.

If we go with `hs.get_module_api()` to move forwards now, it doesn't stop us from making a holistic change in the future.

anoadragon453 replied:
I've decided to go with `hs.get_module_api()` as there's already enough complication with this change.

MadLittleMods commented on s3_storage_provider.py:
```diff
@@ -127,34 +115,29 @@ def _get_s3_client(self):
 
-        self._s3_pool.callInThread(_get_file)
-        return make_deferred_yieldable(d)
+        run_in_background(self._s3_pool.callInThread, _get_file)
```
I think we should use the same `ModuleAPI.defer_to_thread(...)` pattern mentioned above

anoadragon453 committed:
Source `ModuleApi` from `hs` and use `defer_to_thread`

We also make `store_file` and `fetch` async, as they are async in the
base class. This also simplifies the implementation.

We could go through and convert the whole module from deferreds to
async, but that should be done separately.

anoadragon453 commented:
https://github.com/matrix-org/synapse-s3-storage-provider/pull/134/commits/190686c44ec602a762b49d570f275a2f657f9b01 makes use of `ModuleApi.defer_to_thread`, which allows modules to easily run a function on a separate thread. However, you must use the default threadpool.

This module previously used its own threadpool (`self._s3_pool`), and even had an option to configure the size (`threadpool_size`, default 40). In previous issues, we've suggested users configure boto3's connection count to match the configured threadpool size: https://github.com/matrix-org/synapse-s3-storage-provider/issues/117

The default threadpool size in Twisted is 10. I worry that switching to Synapse's default threadpool will hurt the performance of this module, with no way for users to configure it.
 
Perhaps we should additionally expose [`defer_to_threadpool`](https://github.com/element-hq/synapse/blob/5ce4f8916e7b323dd332caa3b1f214e82523d957/synapse/logging/context.py#L1043-L1101) to modules, allowing modules to specify a threadpool? Though this would break compatibility with older versions of Synapse. We could check if the method existed on `ModuleApi` and use the default threadpool if not.

anoadragon453 committed:
Merge branch 'main' into anoa/stop_using_logging_context_directly

MadLittleMods commented on s3_storage_provider.py:
```diff
@@ -124,37 +110,31 @@ def _get_s3_client(self):
-            s3_download_task(
-                self._get_s3_client(), self.bucket, self.prefix + path, self.extra_args, d, logcontext
-            )
+        await self._module_api.defer_to_thread(
```
> This module previously used its own threadpool (`self._s3_pool`), and even had an option to configure the size (`threadpool_size`, default 40). In previous issues, we've suggested users configure boto3's connection count to match the configured threadpool size: https://github.com/matrix-org/synapse-s3-storage-provider/issues/117
> 
> The default threadpool size in Twisted is 10. I worry that switching to Synapse's default threadpool will hurt the performance of this module, with no way for users to configure it.
>
> *-- @anoadragon453, https://github.com/matrix-org/synapse-s3-storage-provider/pull/134#issuecomment-3360286233*

Can we recommend people increase the size of the the default Twisted threadpool?

Is there any difference in having a separate threadpool?

MadLittleMods replied:
*Discussed in [`#synapse-dev:matrix.org`](https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$RohfJV3YnLDRZalwhLN4MJQ9XBTintYMmdUs-zWqQV4?via=matrix.org&via=node.marinchik.ink&via=element.io),*

Here is why a separate threadpool is important:

> I'd be worried about s3 monopolising the thread pool and blocking other operations (like DNS lookups)
>
> *-- @richvdh*

As @anoadragon453 points out, [DNS lookups](
https://github.com/element-hq/synapse/blob/1c093509ceb04ee8ce0eb6a408b76b0fda3ac87c/synapse/app/_base.py#L558-L567) have their own threadpool already but the point still stands generally.

anoadragon453 replied:
> Can we recommend people increase the size of the the default Twisted threadpool?

We don't currently have such an option in Synapse, but it would be better than nothing.

Perhaps we should try deploying the module with the default threadpool on matrix.org and see if performance suffers? I'm just worried that if we go ahead with not adding any way to configure the threadpool size, yet require people to upgrade this module to use the latest Synapse, they could be stuck between a rock and a hard place.

Otherwise, I think:

> Perhaps we should additionally expose [`defer_to_threadpool`](https://github.com/element-hq/synapse/blob/5ce4f8916e7b323dd332caa3b1f214e82523d957/synapse/logging/context.py#L1043-L1101) to modules, allowing modules to specify a threadpool? Though this would break compatibility with older versions of Synapse. We could check if the method existed on `ModuleApi` and use the default threadpool if not.

May be the way to go.

anoadragon453 replied:
> Can we recommend people increase the size of the the default Twisted threadpool?

The size of the default Twisted threadpool can only be increased through code.

We don't provide a configuration option which tweaks this currently, so users are unable to increase the size of the default threadpool.

---

Above I suggested exposing X to modules. Another alternative is to add an argument to the already-exposed [`defer_to_thread`](https://github.com/element-hq/synapse/blob/ee9768c31fce76a3b46aa50a2fb3c51f52dd641e/synapse/module_api/__init__.py#L1716-L1734) `ModuleApi` method to allow specifying a threadpool, defaulting to the default Twisted threadpool if not provided. It would then use `defer_to_threadpool` under the hood instead of `defer_to_thread`.

MadLittleMods replied:
Sounds workable :+1:

anoadragon453 replied:
I've tried the latter approach in https://github.com/element-hq/synapse/pull/19032.

erikjohnston commented:
FYI I tried this branch on jki.re with latest develop and no new media is being downloaded. It seems to fail with:

```
2025-10-03 10:02:43,349 - twisted - 278 - CRITICAL - sentinel - 
Traceback (most recent call last):
  File "/home/erikj/.virtualenvs/synapse311/lib/python3.11/site-packages/s3_storage_provider.py", line 254, in _stream_to_producer
    raise Exception("Timed out waiting to resume")
Exception: Timed out waiting to resume
```

anoadragon453 committed:
Use `defer_to_threadpool`

As introduced by https://github.com/element-hq/synapse/pull/19032.

anoadragon453 committed:
Don't `await` deferred before passing it back to Synapse

Discovered by @erikj:

> the thread is waiting for synapse to use the S3Responder, but the responder isn't returned to Synapse until the thread is finished

anoadragon453 committed:
await on the correct thing

> the `d` we pass in to `s3_download_task` gets resolved once we connect to S3, and the thread is concluded only once we finish the download.

anoadragon453 committed:
Use `run_in_background` on `defer_to_threadpool`

anoadragon453 committed:
Update documentation to mention dangling coroutines

anoadragon453 commented:
Thanks for the reviews both!

My theory as to why this PR passed the integration test CI, but failed on @erikjohnston's homeserver, is that the integration tests [do not delete](https://github.com/matrix-org/synapse-s3-storage-provider/blob/anoa/stop_using_logging_context_directly/.github/workflows/test.yml#L59-L64) the media from Synapse's local storage after it's uploaded. Thus we don't end up actually fetching the media from the minIO S3 storage upon download.

I'll have a look at modifying the tests to actually do that.

MadLittleMods commented on s3_storage_provider.py:
```diff
@@ -124,37 +120,45 @@ def _get_s3_client(self):
 
-        self._s3_pool.callInThread(_get_file)
-        return make_deferred_yieldable(d)
+        # DO await on `d`, as it will resolve once a connection to S3 has been
+        # opened. We only want to return to Synapse once we can start streaming
+        # chunks.
```
Cross-linking [internal discussion](https://matrix.to/#/!SGNQGPGUwtcPBUotTL:matrix.org/$w6oHnf4PL9tjBsvo_Nz9UNE3yYaMSrfAzYwUAOokdRA?via=jki.re&via=matrix.org&via=element.io) where the cause of the `Exception: Timed out waiting to resume` was figured out.

spantaleev commented:
Using `defer_to_threadpool` makes this (and `s3-storage-provider=v1.6.0`, which includes this patch) not compatible with Synapse `<1.140.0`, because `defer_to_threadpool` was only introduced (by [this patch](https://github.com/element-hq/synapse/commit/e3344dc0c3a11b0bb661657da51655a0b02a4bfc)) in Synapse v1.140.0.

While Synapse [v1.140.0rc1](https://github.com/element-hq/synapse/releases/tag/v1.140.0rc1) warns about `s3-storage-provider=v1.6.0` being required to run that Synapse version, we probably also need a similar warning for `s3-storage-provider`: you can't run `s3-storage-provider=v1.6.0` with older Synapse versions.

Users on the current stable Synapse release (v1.139.0) are better off staying with `s3-storage-provider=v1.5.0`.

anoadragon453 commented:
@spantaleev thanks for raising! We completely forgot to signal that. I've added a warning to the top of the latest release notes: https://github.com/matrix-org/synapse-s3-storage-provider/releases/tag/v1.6.0.

Synapse 1.140.0 is expected to be released tomorrow (as long as no other regressions are found).
